### ISSUE TITLE
多 GPU 训练在第二个 epoch 卡在 loss.backward()

### ISSUE THREAD

**Issue by panpan-wu** [wenet-e2e/wenet#795]
用 strace 追踪，错误信息：futex(0x55dcd81e08c8, FUTEX_WAIT_PRIVATE, 0, NULL
感觉是数据不平衡导致的。

**Comment by panpan-wu** [comment]
训练时用的是动态 batch，batch_type: 'dynamic'。当动态 batch，梯度累积和多 GPU 一起用时，得到的梯度的期望与梯度的真实值是否一致呢？

**Comment by panpan-wu** [comment]
换成 static batch 后不会卡了。

**Comment by kobenaxie** [comment]
dynamic有同样的问题，换成static容易OOM

**Comment by joeychsu** [comment]
I have same issues.
if batch_type -> dynamic will get stuck on epoch 1
if batch_type -> static will get OOM easily

**Comment by panpan-wu** [comment]
train.sh
```bash
#!/bin/bash

# export CUDA_VISIBLE_DEVICES="0,1,2,3"
export CUDA_VISIBLE_DEVICES="0,1,2,3"

dir=exp/conformer
cmvn=true
# checkpoint=exp/20210815_unified_conformer_server/final.pt
# checkpoint=exp/conformer/init.pt
# bpemodel=conf/train_960_unigram5000.model

# The NCCL_SOCKET_IFNAME variable specifies which IP interface to use for nccl
# communication. More details can be found in
# https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html
# export NCCL_SOCKET_IFNAME=ens4f1
# The num of nodes or machines used for multi-machine training
# Default 1 for single machine/node
# NFS will be needed if you want run multi-machine training
num_nodes=1
# The rank of each node or machine, range from 0 to num_nodes -1
# The first node/machine sets node_rank 0, the second one sets node_rank 1
# the third one set node_rank 2, and so on. Default 0
node_rank=0

# Training
mkdir -p $dir
# INIT_FILE=$dir/ddp_init
# You had better rm it manually before you start run.sh on first node.
# rm -f $INIT_FILE # delete old one before starting
# init_method=file://$(readlink -f $INIT_FILE)
init_method=tcp://127.0.0.1:9000
echo "$0: init method is $init_method"
num_gpus=$(echo $CUDA_VISIBLE_DEVICES | awk -F "," '{print NF}')
# Use "nccl" if it works, otherwise use "gloo"
dist_backend="nccl"
# The total number of processes/gpus, so that the master knows
# how many workers to wait for.
# More details about ddp can be found in
# https://pytorch.org/tutorials/intermediate/dist_tuto.html
world_size=`expr $num_gpus \* $num_nodes`
echo "total gpus is: $world_size"
cmvn_opts=
$cmvn && cp data/all/train/global_cmvn $dir/global_cmvn
$cmvn && cmvn_opts="--cmvn ${dir}/global_cmvn"
# train.py will write $train_config to $dir/train.yaml with model input
# and output dimension, train.yaml will be used for inference or model
# export later
start_epoch=27
end_epoch=127
for epoch in `seq $start_epoch $end_epoch`; do
    if [ $epoch -eq -1 ]; then
        checkpoint=exp/conformer.prev/89.pt
    else
        checkpoint=exp/conformer/${epoch}.pt
    fi
    echo "CHECKPOINT: $checkpoint"
    for ((i = 0; i < $num_gpus; ++i)); do
    {
        gpu_id=$(echo $CUDA_VISIBLE_DEVICES | cut -d',' -f$[$i+1])
        # Rank of each gpu/process used for knowing whether it is
        # the master of a worker.
        rank=`expr $node_rank \* $num_gpus + $i`

        python /opt/wenet/wenet/bin/train.py --gpu $gpu_id \
            --config conf/train_unified_conformer.yaml \
            --data_type raw \
            --symbol_table data/all/train/words.txt \
            --train_data data/all/train/data.list \
            --cv_data data/all/dev/data.list \
            --checkpoint $checkpoint \
            --model_dir $dir \
            --ddp.init_method $init_method \
            --ddp.world_size $world_size \
            --ddp.rank $rank \
            --ddp.dist_backend $dist_backend \
            --num_workers 4 \
            $cmvn_opts \
            --pin_memory
    } &
    done
    wait
done
```

**Comment by panpan-wu** [comment]
batch conf 配置：
```yaml
    batch_conf:
        # batch_type: 'static' # static or dynamic
        batch_type: 'dynamic' # static or dynamic
        # batch_size: 4
        max_frames_in_batch: 20000
```

**Comment by kli017** [comment]
@panpan-wu 所以dynamic要一直手动改epoch吗，捂脸

**Comment by panpan-wu** [comment]
我对代码做了一些改动，不会卡了。

**Comment by panpan-wu** [comment]
> @panpan-wu 所以dynamic要一直手动改epoch吗，捂脸

基本是吧，不过使用 shell 脚本改的：
for epoch in `seq $start_epoch $end_epoch`; do
    ...
done

后来我改了一些代码，不会卡了，见评论。

**Comment by xingchensong** [comment]
torch 的 join() 以及 cv的static 可以解决这个问题，see https://github.com/wenet-e2e/wenet/issues/842 and https://github.com/wenet-e2e/wenet/pull/903
